apps/hermes/server/src/state/aggregate.rs

```diff
@@ -568,6 +616,145 @@ where
     })
 }
 
+async fn get_verified_twaps_with_update_data<S>(
+    state: &S,
+    price_ids: &[PriceIdentifier],
+    start_time: RequestTime,
+    end_time: RequestTime,
+) -> Result<TwapsWithUpdateData>
+where
+    S: Cache,
+{
+    let mut calculated_twaps = Vec::new();
+    let mut all_update_data = Vec::new();
+
+    for price_id in price_ids {
+        let start_messages = state
+            .fetch_message_states(
+                vec![price_id.to_bytes()],
+                start_time.clone(),
+                MessageStateFilter::Only(MessageType::TwapMessage),
+            )
+            .await?;
+
+        let end_messages = state
+            .fetch_message_states(
+                vec![price_id.to_bytes()],
+                end_time.clone(),
+                MessageStateFilter::Only(MessageType::TwapMessage),
+            )
+            .await?;
+
+        if start_messages.is_empty() || end_messages.is_empty() {
+            continue;
+        }
+
+        let start_message = &start_messages[0];
+        let end_message = &end_messages[0];
+
+        if let (Message::TwapMessage(start_twap), Message::TwapMessage(end_twap)) =
+            (&start_message.message, &end_message.message)
+        {
+            match calculate_twap(start_twap, end_twap) {
+                Ok(twap_price) => {
+                    // down_slots_ratio describes the % of slots where the network was down
+                    // over the TWAP window. A value closer to zero indicates higher confidence.
+                    let total_slots = end_message.slot - start_message.slot;
```

Comment:
does this need to be `publish_slot` ?

Reply:
not sure of the difference tbh :) updated